media/windows/ECC-Meraki-get-client-data/ECC-Meraki-get-client-data.py

```diff
@@ -0,0 +1,1205 @@
+
+    db_insert_sql_str += '''?)'''
+
+    logger.debug(F"SQL INSERT statement:  '{db_insert_sql_str}'")
+
+    return db_insert_sql_str
+
+
+def create_database_record(conn, db_table, network_name, network_id, client, values_dict):
+    """
+        This routine will attempt to write a record to the passed table with the passed
+        list of record names / values.  If successful, the routine will return True, else False.
+            Written by DK Fowler ... 22-Jan-2020
+    :param conn:            database connection (may exist already, or be set to None)
+    :param db_table:        table name to which a record write is attempted
+    :param network_name:    network name
+    :param network_id:      network ID
+    :param client           client dictionary containing field name: value
+    :param values_dict:     dictionary of field names: field values for the record
+    :return:                True if table created successfully; otherwise, False
+    """
+
+    # First create a SQL string for INSERTing the record into the passed table
+    sql_insert = construct_insert_table_sql(db_table, values_dict.keys())
+
+    # Get the current date/time in UTC for the record INSERT
+    recordWrittenUTC = datetime.utcnow()
+
+    logger.debug(F"......date written set to: {datetime.strftime(recordWrittenUTC, '%Y-%m-%d %H:%M:%S.%f')}")
+    cur = conn.cursor()  # Get a cursor for the db connection
+    # Make sure we have a database connection
+    if conn is not None:
+        logger.debug(F"Attempting record insert...table:  {db_table}")
+        try:
+            val_list = []
+            date_written_str = datetime.strftime(recordWrittenUTC, "%Y-%m-%d %H:%M:%S.%f")
+            val_list.insert(0, date_written_str)  # insert common key values in the beginning of the value list
+            val_list.insert(1, network_name)
+            val_list.insert(2, network_id)
+            idx = 3
+            # for db_key, db_val in client.items(): # removed 8-Sep-2021, DKF
+            for db_val in client.values():
+                # Usage in sent/received kilobytes is included in the client record as an embedded dictionary;
+                # break this into separate send/received fields for ease in analysis.
+                # *** No longer required as of v03.70, as passed client has now been reordered
+                # *** to match the db table structure.
+                #       Removed 08-Sep-2021...DK Fowler
+                """
+                if db_key == 'usage':
+                    val_list.insert(idx, client['usage']['sent'])
+                    idx += 1
+                    val_list.insert(idx, client['usage']['recv'])
+                    idx += 1
+                else:
+                """
+                val_list.insert(idx, db_val)
+                idx += 1
+
+            # print(F"Values used in INSERT:  {val_list}")
+            logger.debug(F"...field values {val_list}")
+            cur.execute(sql_insert, val_list)
```

Comment:
Instead of using `?` in the SQL string and a list of values, it's a little safer to use named fields and a dictionary.  That way, the ordering of the fields is less important -- particularly if they shift around over time.  Here's an example from the stewardship code -- `cookies` is the SQLite cursor:

```python
# Insert the URL in the cookies database                                                                                                                                 
query       = ("INSERT INTO cookies "
                "(cookie,fid,url) "
                "VALUES (:cookie, :fid, :url);")
values      = {
    "cookie" : cookie,
    "fid"    : fid,
    "url"    : url_escaped,
}
cookies.execute(query, values)
```

You can ignore the `id` and `recordWrittenUTC` fields since they'll be written by SQLite3 automatically; you can just iterate over the field names from `expected_col_names` to generate both the SQL string and the values dictionary:

```python
sql = 'INSERT INTO data('
sql += ','.join(expected_col_names.keys())
sql += ') VALUES ('
names = [f':{name}' for name in expected_col_names.keys()]
sql += ','.join(names)
sql += ');'
```

(there's probably a more clever way to do that than making a `names` list, but you get the idea)

This makes a string like this:

```
INSERT INTO data(networkName,networkID,...) VALUES (:networkName,:networkID,...);
```

Then just have a `values` dictionary with keys that are the same name as the field names from `expected_col_names`, and you can `foo.execute(sql, values)`.

Reply:
Ah, I like that technique!  It does solve the ordering problem.  Hadn't run across it in other code examples...I'll note that for future work.  Probably not going to bother changing the existing code at this point though.